### PR TITLE
[MOT] add tricks for fairmot

### PR DESCRIPTION
**Pull request by nemonameless**
1. add SGE attention
2. add iou head with GIoU loss

### CODE REVIEW

File: ppdet/modeling/heads/centernet_head.py [hunk 1]
```diff
@@ -111,13 +120,45 @@ def __init__(self,
...
-            ConvLayer(
-                in_channels, head_planes, kernel_size=3, padding=1, bias=True),
-            nn.ReLU(),
-            ConvLayer(
-                head_planes, 2, kernel_size=1, stride=1, padding=0, bias=True))
-        self.focal_loss = CTFocalLoss()
+        self.size_loss = size_loss
+
+        # offset head (optinal)
+        self.offset_weight = offset_weight
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py]: offset_weight已经在前面赋给self.weights了，这里不需要再重复赋值

File: ppdet/modeling/heads/centernet_head.py [hunk 2]
```diff
@@ -128,23 +169,32 @@ def from_config(cls, cfg, input_shape):
     def forward(self, feat, inputs):
         heatmap = self.heatmap(feat)
         size = self.size(feat)
-        offset = self.offset(feat)
+        offset = self.offset(feat) if self.offset_weight else None
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py]: offset = self.offset(feat) if hasattr(self, 'offset_weight') else None

File: ppdet/modeling/heads/centernet_head.py [hunk 3]
```diff
@@ -162,32 +212,86 @@ def get_loss(self, heatmap, size, offset, weights, inputs):
...
+            size_loss = size_loss / (pos_num + 1e-4)
+        elif self.size_loss == 'giou':
+            size_target = inputs['bbox_xys']
+            size_target.stop_gradient = True
+            centers_x = (size_target[:, :, 0:1] + size_target[:, :, 2:3]) / 2.0
+            centers_y = (size_target[:, :, 1:2] + size_target[:, :, 3:4]) / 2.0
+            x1 = centers_x - pos_size[:, :, 0:1]
+            y1 = centers_y - pos_size[:, :, 1:2]
+            x2 = centers_x + pos_size[:, :, 2:3]
+            y2 = centers_y + pos_size[:, :, 3:4]
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py:212]: 用真值的中心点作为预测值的中心点，逻辑上讲不通。要么就是预测框自己的中心点

File: ppdet/modeling/heads/centernet_head.py [hunk 4]
```diff
@@ -162,32 +212,86 @@ def get_loss(self, heatmap, size, offset, weights, inputs):
...
+            iou_mask.stop_gradient = True
+            gt_bbox_xys = inputs['bbox_xys']
+            gt_bbox_xys.stop_gradient = True
+            centers_x = (gt_bbox_xys[:, :, 0:1] + gt_bbox_xys[:, :, 2:3]) / 2.0
+            centers_y = (gt_bbox_xys[:, :, 1:2] + gt_bbox_xys[:, :, 3:4]) / 2.0
+            x1 = centers_x - pos_size[:, :, 0:1]
+            y1 = centers_y - pos_size[:, :, 1:2]
+            x2 = centers_x + pos_size[:, :, 2:3]
+            y2 = centers_y + pos_size[:, :, 3:4]
+            pred_boxes = paddle.concat([x1, y1, x2, y2], axis=-1)
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py:257]: 同上

File: ppdet/modeling/necks/centernet_fpn.py [hunk 5]
```diff
@@ -16,12 +16,81 @@
...
 import paddle.nn as nn
+import paddle.nn.functional as F
 from paddle.nn.initializer import KaimingUniform
 from ppdet.core.workspace import register, serializable
 from ppdet.modeling.layers import ConvNormLayer
 from ..shape_spec import ShapeSpec
 
 
+# SGE attention
+class BasicConv(nn.Layer):
```
Comment (FlyingQianMM) [inline ppdet/modeling/necks/centernet_fpn.py:31]: BasicConv可以复用ppdet.modeling.layers里的ConvNormLayer么

File: ppdet/modeling/heads/centernet_head.py [hunk 6]
```diff
@@ -162,32 +212,86 @@ def get_loss(self, heatmap, size, offset, weights, inputs):
...
+                pred_boxes * size_mask,
+                size_target * size_mask,
+                iou_weight=size_mask,
+                loc_reweight=None)
+            size_loss = size_loss / (pos_num + 1e-4)
 
-        det_loss = weights['heatmap'] * heatmap_loss + weights[
-            'size'] * size_loss + weights['offset'] * offset_loss
+        # offset head loss: L1 loss
+        if self.offset_weight:
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py]: if offset is not None:

File: ppdet/modeling/heads/centernet_head.py [hunk 7]
```diff
@@ -162,32 +212,86 @@ def get_loss(self, heatmap, size, offset, weights, inputs):
...
+            offset_loss = F.l1_loss(
+                pos_offset * offset_mask,
+                offset_target * offset_mask,
+                reduction='sum')
+            offset_loss = offset_loss / (pos_num + 1e-4)
 
-        return {
-            'det_loss': det_loss,
+        # iou head loss: GIoU loss
+        if self.iou_weight:
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py]: if iou is not None:

File: ppdet/modeling/heads/centernet_head.py [hunk 8]
```diff
@@ -162,32 +212,86 @@ def get_loss(self, heatmap, size, offset, weights, inputs):
...
+            iou_loss = iou_loss / (pos_num + 1e-4)
+
+        losses = {
             'heatmap_loss': heatmap_loss,
             'size_loss': size_loss,
-            'offset_loss': offset_loss
         }
+        det_loss = weights['heatmap'] * heatmap_loss + weights[
+            'size'] * size_loss
+        if self.offset_weight:
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py]: if offset is not None:

File: ppdet/modeling/necks/centernet_fpn.py (hunk 5 above)
Comment (nemonameless) [reply ppdet/modeling/necks/centernet_fpn.py:31]: 功能上类似，但是细节待替换后实验看效果

File: ppdet/modeling/heads/centernet_head.py (hunk 3 above)
Comment (nemonameless) [reply ppdet/modeling/heads/centernet_head.py:212]: 最终MOTA最高的模型未使用giou loss

File: ppdet/modeling/necks/centernet_fpn.py (hunk 5 above)
Comment (nemonameless) [reply ppdet/modeling/necks/centernet_fpn.py:31]: 建议暂不替换

**Review by jerrywgz — APPROVED** [review]

**Review by qingqing01 — APPROVED** [review]
